File: src/precalculus_polar_coordinates_hard.py
"""
precalculus - polar_coordinates (hard)
"""

from problem_utils import *

ALL_ANGLES = [
    0, pi/6, pi/4, pi/3, pi/2,
    2*pi/3, 3*pi/4, 5*pi/6, pi,
    7*pi/6, 5*pi/4, 4*pi/3, 3*pi/2,
    5*pi/3, 7*pi/4, 11*pi/6,
]


def generate():
    problem_type = choice([
        'convert_to_rect_circle',
        'convert_to_rect_line',
        'convert_to_polar',
        'intersection_circle_circle',
        'intersection_circle_line',
        'area_cardioid',
        'area_rose',
        'arc_length_circle',
    ])

    if problem_type == 'convert_to_rect_circle':
        # r = 2a cos(theta) or r = 2b sin(theta)
        a = choice([1, 2, 3, 4, 5])
        use_cos = choice([True, False])
        if use_cos:
            polar_eq = f"r = {2*a}\\cos\\theta"
            rect_ans = Eq((x - a)**2 + y**2, a**2)
            sol = steps(
                f"Multiply by $r$: $r^2 = {2*a}r\\cos\\theta$",
                f"$x^2+y^2 = {2*a}x$",
                f"$(x-{a})^2+y^2={a**2}$  (circle, center $({a},0)$, radius ${a}$)"
            )
        else:
            polar_eq = f"r = {2*a}\\sin\\theta"
            rect_ans = Eq(x**2 + (y - a)**2, a**2)
            sol = steps(
                f"Multiply by $r$: $r^2 = {2*a}r\\sin\\theta$",
                f"$x^2+y^2 = {2*a}y$",
                f"$x^2+(y-{a})^2={a**2}$  (circle, center $(0,{a})$, radius ${a}$)"
            )
        return problem(
            question=f"Convert ${polar_eq}$ to rectangular form.",
            answer=rect_ans,
            difficulty=(1600, 1750),
            topic="precalculus/polar_coordinates",
            solution=sol,
            answer_type="equation"
        )

    elif problem_type == 'convert_to_rect_line':
        variant = choice(['vertical', 'horizontal', 'slant_pi4', 'normal'])
        if variant == 'vertical':
            a = randint(1, 8)
            sign = choice([-1, 1])
            a_s = sign * a
            polar_eq = f"r\\cos\\theta = {a_s}"
            rect_ans = Eq(x, a_s)
            sol = steps("$r\\cos\\theta = x$", f"$x = {a_s}$")
        elif variant == 'horizontal':
            a = randint(1, 8)
            sign = choice([-1, 1])
            a_s = sign * a
            polar_eq = f"r\\sin\\theta = {a_s}"
            rect_ans = Eq(y, a_s)
            sol = steps("$r\\sin\\theta = y$", f"$y = {a_s}$")
        elif variant == 'slant_pi4':
            a = randint(1, 6)
            sign = choice([-1, 1])
            a_s = sign * a
            polar_eq = f"r\\cos\\!\\left(\\theta - \\dfrac{{\\pi}}{{4}}\\right) = {a_s}"
            rect_ans = Eq(x + y, a_s * sqrt(2))
            sol = steps(
                f"Expand: $r\\!\\left(\\cos\\theta\\cos\\tfrac{{\\pi}}{{4}}+\\sin\\theta\\sin\\tfrac{{\\pi}}{{4}}\\right)={a_s}$",
                f"$\\tfrac{{\\sqrt2}}{{2}}(r\\cos\\theta+r\\sin\\theta)={a_s}$",
                f"$x+y = {a_s}\\sqrt{{2}} = {latex(a_s*sqrt(2))}$"
            )
        else:
            a = randint(1, 6)
            sign = choice([-1, 1])
            a_s = sign * a
            polar_eq = f"r\\cos\\!\\left(\\theta - \\dfrac{{\\pi}}{{6}}\\right) = {a_s}"
            rect_ans = Eq(sqrt(3)*x + y, 2*a_s)
            sol = steps(
                f"Expand: $r\\cos\\theta\\cos\\tfrac{{\\pi}}{{6}}+r\\sin\\theta\\sin\\tfrac{{\\pi}}{{6}}={a_s}$",
                f"$\\tfrac{{\\sqrt3}}{{2}}x+\\tfrac{{1}}{{2}}y={a_s}$",
                f"$\\sqrt{{3}}\\,x+y={2*a_s}$"
            )
        return problem(
            question=f"Convert ${polar_eq}$ to rectangular form.",
            answer=rect_ans,
            difficulty=(1650, 1800),
            topic="precalculus/polar_coordinates",
            solution=sol,
            answer_type="equation"
        )

    elif problem_type == 'convert_to_polar':
        rect_type = choice(['circle_origin', 'circle_shifted_x', 'circle_shifted_y', 'line_y_eq_mx', 'parabola'])

        if rect_type == 'circle_origin':
            r_val = randint(2, 9)
            rect_eq = f"x^2 + y^2 = {r_val**2}"
            polar_ans = Eq(symbols('r'), r_val)
            sol = steps(f"$r^2 = x^2+y^2 = {r_val**2}$", f"$r = {r_val}$")
        elif rect_type == 'circle_shifted_x':
            a = randint(1, 5)
            rect_eq = f"(x-{a})^2 + y^2 = {a**2}"
            polar_ans = Eq(symbols('r'), 2*a*cos(symbols('theta')))
            sol = steps(
                f"Expand: $x^2-{2*a}x+{a**2}+y^2={a**2}$",
                f"$x^2+y^2={2*a}x \\Rightarrow r^2={2*a}r\\cos\\theta$",
                f"$r = {2*a}\\cos\\theta$"
            )
        elif rect_type == 'circle_shifted_y':
            a = randint(1, 5)
            rect_eq = f"x^2 + (y-{a})^2 = {a**2}"
            polar_ans = Eq(symbols('r'), 2*a*sin(symbols('theta')))
            sol = steps(
                f"$x^2+y^2={2*a}y \\Rightarrow r^2={2*a}r\\sin\\theta$",
                f"$r = {2*a}\\sin\\theta$"
            )
        elif rect_type == 'line_y_eq_mx':
            m = choice([1, 2, 3, -1, -2])
            b_v = choice([1, 2, 3, -1, -2])
            rect_eq = f"y = {m}x + {b_v}"
            theta_sym = symbols('theta')
            r_sym = symbols('r')
            polar_ans = Eq(r_sym, b_v / (sin(theta_sym) - m*cos(theta_sym)))
            sol = steps(
                f"Substitute $x=r\\cos\\theta$, $y=r\\sin\\theta$:",
                f"$r\\sin\\theta = {m}r\\cos\\theta + {b_v}$",
                f"$r(\\sin\\theta - {m}\\cos\\theta) = {b_v}$",
                f"$r = \\dfrac{{{b_v}}}{{\\sin\\theta - {m}\\cos\\theta}}$"
            )
        else:
            # y^2 = 4x  parabola
            a = randint(1, 4)
            rect_eq = f"y^2 = {4*a}x"
            theta_sym = symbols('theta')
            r_sym = symbols('r')
            # r^2 sin^2 = 4a r cos => r = 4a cos / sin^2
            polar_ans = Eq(r_sym, 4*a*cos(theta_sym)/sin(theta_sym)**2)
            sol = steps(
                f"Substitute $x=r\\cos\\theta$, $y=r\\sin\\theta$:",
                f"$r^2\\sin^2\\theta = {4*a}r\\cos\\theta$",
                f"$r = \\dfrac{{{4*a}\\cos\\theta}}{{\\sin^2\\theta}}$"
            )
        return problem(
            question=f"Convert ${rect_eq}$ to polar form.",
            answer=polar_ans,
            difficulty=(1650, 1800),
            topic="precalculus/polar_coordinates",
            solution=sol,
            answer_type="equation"
        )

    elif problem_type == 'intersection_circle_circle':
        # r = a  and  r = b cos(theta), a < b
        a = randint(2, 4)
        b = randint(a + 1, a + 4)
        ratio = Rational(a, b)
        theta_pos = simplify(acos(ratio))
        theta_neg = simplify(-theta_pos)
        polar_ans = FiniteSet((a, theta_pos), (a, theta_neg))
        return problem(
            question=(f"Find the intersection points of $r = {a}$ and $r = {b}\\cos\\theta$ "
                      f"in the form $(r,\\theta)$ with $-\\pi<\\theta\\le\\pi$."),
            answer=polar_ans,
            difficulty=(1700, 1850),
            topic="precalculus/polar_coordinates",
            solution=steps(
                f"Set equal: ${a} = {b}\\cos\\theta$",
                f"$\\cos\\theta = {latex(ratio)}$",
                f"$\\theta = \\pm\\arccos\\!\\left({latex(ratio)}\\right) = \\pm{latex(theta_pos)}$",
                f"Intersection points: $({a}, {latex(theta_pos)})$ and $({a}, {latex(theta_neg)})$"
            ),
            answer_type="set",
            calculator="scientific"
        )

    elif problem_type == 'intersection_circle_line':
        # r = a  and  r = b sin(theta), a < b  (same pattern)
        a = randint(2, 4)
        b = randint(a + 1, a + 4)
        ratio = Rational(a, b)
        theta1 = simplify(asin(ratio))
        theta2 = simplify(pi - theta1)
        polar_ans = FiniteSet((a, theta1), (a, theta2))
        return problem(
            question=(f"Find the intersection points of $r = {a}$ and $r = {b}\\sin\\theta$ "
                      f"with $0\\le\\theta<\\pi$."),
            answer=polar_ans,
            difficulty=(1700, 1850),
            topic="precalculus/polar_coordinates",
            solution=steps(
                f"Set equal: ${a} = {b}\\sin\\theta$",
                f"$\\sin\\theta = {latex(ratio)}$",
                f"$\\theta = {latex(theta1)}$ or $\\theta = {latex(theta2)}$",
                f"Points: $({a},{latex(theta1)})$ and $({a},{latex(theta2)})$"
            ),
            answer_type="set",
            calculator="scientific"
        )

    elif problem_type == 'area_cardioid':
        a = randint(1, 6)
        use_cos = choice([True, False])
        # Area of r = a(1 + cos(theta)) or r = a(1 + sin(theta)) = (3/2)*pi*a^2
        area_ans = Rational(3, 2) * pi * a**2
        func_str = "\\cos" if use_cos else "\\sin"
        return problem(
            question=f"Find the area enclosed by the cardioid $r = {a}(1 + {func_str}\\,\\theta)$.",
            answer=area_ans,
            difficulty=(1750, 1900),
            topic="precalculus/polar_coordinates",
            solution=steps(
                "$A = \\dfrac{1}{2}\\displaystyle\\int_0^{2\\pi} r^2\\,d\\theta$",
                f"$r^2 = {a**2}(1+{func_str}\\,\\theta)^2 = {a**2}\\!\\left(1+2{func_str}\\,\\theta+{func_str}^2\\theta\\right)$",
                f"Using $\\cos^2\\theta=\\tfrac{{1+\\cos(2\\theta)}}{{2}}$ (or $\\sin^2$):",
                f"$A = \\dfrac{{{a**2}}}{{2}}\\cdot 3\\pi = {latex(area_ans)}$"
            )
        )

    elif problem_type == 'area_rose':
        n = choice([2, 3, 4, 5])
        a = randint(1, 5)
        petals = n if n % 2 == 1 else 2 * n
        # Area of one petal of r = a sin(n theta): pi*a^2/(4n) for odd n, pi*a^2/(4n) for even
        # Total area: petals * (pi*a^2)/(4n)  ... formula: pi*a^2/2 for odd n, pi*a^2/2 for even n
        # For r = a*cos(n*theta): total area = pi*a^2/2
        total_area = Rational(1, 2) * pi * a**2
        func_str = choice(['\\cos', '\\sin'])
        return problem(
            question=f"Find the total area enclosed by the rose $r = {a}{func_str}({n}\\theta)$.",
            answer=total_area,
            difficulty=(1800, 1950),
            topic="precalculus/polar_coordinates",
            solution=steps(
                f"The rose has ${petals}$ petals",
                f"Area of one petal: $\\dfrac{{1}}{{2}}\\displaystyle\\int_0^{{\\pi/{n}}} {a**2}{func_str}^2({n}\\theta)\\,d\\theta = \\dfrac{{\\pi {a**2}}}{{4n}} \\cdot 2 = \\dfrac{{\\pi {a**2}}}{{{2*n}}}$",
                f"Total area: ${petals} \\times \\dfrac{{\\pi {a**2}}}{{{2*n}}} = {latex(total_area)}$"
            )
        )

    else:
        # Arc length of r = a (circle)
        a = randint(2, 8)
        theta1 = 0
        theta2 = choice([pi/2, pi, 3*pi/2, 2*pi])
        arc = simplify(a * theta2)
        frac_str = {pi/2: "\\pi/2", pi: "\\pi", 3*pi/2: "3\\pi/2", 2*pi: "2\\pi"}[theta2]
        return problem(
            question=f"Find the arc length of $r = {a}$ from $\\theta = 0$ to $\\theta = {frac_str}$.",
            answer=arc,
            difficulty=(1700, 1850),
            topic="precalculus/polar_coordinates",
            solution=steps(
                "$L = \\displaystyle\\int_\\alpha^\\beta \\sqrt{r^2 + \\left(\\tfrac{dr}{d\\theta}\\right)^2}\\,d\\theta$",
                f"For $r={a}$ (constant): $\\tfrac{{dr}}{{d\\theta}}=0$",
                f"$L = \\displaystyle\\int_0^{{{frac_str}}} {a}\\,d\\theta = {a}\\cdot{frac_str} = {latex(arc)}$"
            )
        )


emit(generate())

File: src/precalculus_polar_coordinates_hard.jl
# precalculus - polar_coordinates (hard)
# Generated: 2026-03-08T20:33:26.176354

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script r θ begin
    set_topic!("precalculus/polar_coordinates")
    
    problem_type = choice([
        :polar_to_cartesian_conversion,
        :cartesian_to_polar_conversion,
        :polar_curve_intersection,
        :polar_area_setup,
        :polar_distance,
    ])
    
    if problem_type == :polar_to_cartesian_conversion
        # Convert polar equation to Cartesian
        subtype = choice([:circle, :line, :parabola])
        
        if subtype == :circle
            a = nonzero(-8, 8)
            # r = 2a cos(θ) or r = 2a sin(θ)
            if choice([true, false])
                polar_eq = "r = $(2*a)\\cos(\\theta)"
                # x² + y² = 2ax  →  (x-a)² + y² = a²
                cart_eq = "(x - $(a))^2 + y^2 = $(a^2)"
                center = (a, 0)
                radius = abs(a)
            else
                polar_eq = "r = $(2*a)\\sin(\\theta)"
                # x² + y² = 2ay  →  x² + (y-a)² = a²
                cart_eq = "x^2 + (y - $(a))^2 = $(a^2)"
                center = (0, a)
                radius = abs(a)
            end
            
            problem(
                question="Convert the polar equation \\\$$(polar_eq)\\\$ to Cartesian form and identify the curve.",
                answer=cart_eq,
                difficulty=(1800, 2000),
                solution=steps(
                    "Given: \\\$$(polar_eq)\\\$",
                    "Multiply both sides by \\\$r\\\$: \\\$r^2 = $(2*a)r\\cos(\\theta)\\\$ or \\\$r^2 = $(2*a)r\\sin(\\theta)\\\$",
                    "Substitute \\\$r^2 = x^2 + y^2\\\$ and \\\$r\\cos(\\theta) = x\\\$ or \\\$r\\sin(\\theta) = y\\\$",
                    "Simplify: \\\$$(cart_eq)\\\$",
                    "This is a circle with center $(center) and radius $(radius)",
                ),
                time=180,
            )
            
        elseif subtype == :line
            m = nonzero(-5, 5)
            b = nonzero(-8, 8)
            # Line y = mx + b in polar: r sin(θ) = m r cos(θ) + b
            # r(sin(θ) - m cos(θ)) = b
            # r = b / (sin(θ) - m cos(θ))
            
            polar_eq = "r = \\frac{$(b)}{\\sin(\\theta) - $(m)\\cos(\\theta)}"
            cart_eq = "y = $(m)x + $(b)"
            
            problem(
                question="Convert the polar equation \\\$$(polar_eq)\\\$ to Cartesian form.",
                answer=cart_eq,
                difficulty=(1900, 2100),
                solution=steps(
                    "Given: \\\$$(polar_eq)\\\$",
                    "Multiply both sides: \\\$r\\sin(\\theta) - $(m)r\\cos(\\theta) = $(b)\\\$",
                    "Substitute \\\$x = r\\cos(\\theta)\\\$ and \\\$y = r\\sin(\\theta)\\\$",
                    "Result: \\\$y - $(m)x = $(b)\\\$",
                    sol("Answer", cart_eq),
                ),
                time=180,
            )
        else # parabola
            p = nonzero(-6, 6)
            # Parabola with focus at origin: r = p/(1 + cos(θ)) or similar
            orientation = choice([:right, :left, :up, :down])
            
            if orientation == :right
                polar_eq = "r = \\frac{$(abs(p))}{1 - \\cos(\\theta)}"
                cart_eq = "y^2 = $(4*abs(p))(x + $(abs(p)))"
            elseif orientation == :left
                polar_eq = "r = \\frac{$(abs(p))}{1 + \\cos(\\theta)}"
                cart_eq = "y^2 = $(-4*abs(p))(x - $(abs(p)))"
            elseif orientation == :up
                polar_eq = "r = \\frac{$(abs(p))}{1 - \\sin(\\theta)}"
                cart_eq = "x^2 = $(4*abs(p))(y + $(abs(p)))"
            else
                polar_eq = "r = \\frac{$(abs(p))}{1 + \\sin(\\theta)}"
                cart_eq = "x^2 = $(-4*abs(p))(y - $(abs(p)))"
            end
            
            problem(
                question="Convert the polar equation \\\$$(polar_eq)\\\$ to Cartesian form.",
                answer=cart_eq,
                difficulty=(2100, 2300),
                solution=steps(
                    "Given: \\\$$(polar_eq)\\\$",
                    "This is a conic section with eccentricity e = 1 (parabola)",
                    "Use \\\$r^2 = x^2 + y^2\\\$ and polar-to-Cartesian conversions",
                    "Multiply and substitute to eliminate \\\$r\\\$ and \\\$\\theta\\\$",
                    sol("Answer", cart_eq),
                ),
                time=240,
            )
        end
        
    elseif problem_type == :cartesian_to_polar_conversion
        # Convert Cartesian to polar
        a = nonzero(-10, 10)
        b = nonzero(-10, 10)
        
        conversion_type = choice([:vertical_line, :horizontal_line, :circle_origin])
        
        if conversion_type == :vertical_line
            cart_eq = "x = $(a)"
            polar_eq = "r\\cos(\\theta) = $(a)"
            simplified = "r = $(a)\\sec(\\theta)"
            
            problem(
                question="Convert the Cartesian equation \\\$$(cart_eq)\\\$ to polar form.",
                answer=simplified,
                difficulty=(1800, 1950),
                solution=steps(
                    "Given: \\\$$(cart_eq)\\\$",
                    "Substitute \\\$x = r\\cos(\\theta)\\\$",
                    sol("Result", polar_eq),
                    sol("Simplified", simplified),
                ),
                time=120,
            )
            
        elseif conversion_type == :horizontal_line
            cart_eq = "y = $(b)"
            polar_eq = "r\\sin(\\theta) = $(b)"
            simplified = "r = $(b)\\csc(\\theta)"
            
            problem(
                question="Convert the Cartesian equation \\\$$(cart_eq)\\\$ to polar form.",
                answer=simplified,
                difficulty=(1800, 1950),
                solution=steps(
                    "Given: \\\$$(cart_eq)\\\$",
                    "Substitute \\\$y = r\\sin(\\theta)\\\$",
                    sol("Result", polar_eq),
                    sol("Simplified", simplified),
                ),
                time=120,
            )
            
        else # circle through origin
            R = abs(nonzero(-12, 12))
            if choice([true, false])
                cart_eq = "x^2 + y^2 = $(R)x"
                polar_eq = "r = $(R)\\cos(\\theta)"
            else
                cart_eq = "x^2 + y^2 = $(R)y"
                polar_eq = "r = $(R)\\sin(\\theta)"
            end
            
            problem(
                question="Convert the Cartesian equation \\\$$(cart_eq)\\\$ to polar form.",
                answer=polar_eq,
                difficulty=(1850, 2000),
                solution=steps(
                    "Given: \\\$$(cart_eq)\\\$",
                    "Substitute \\\$r^2 = x^2 + y^2\\\$, \\\$x = r\\cos(\\theta)\\\$, \\\$y = r\\sin(\\theta)\\\$",
                    "Divide both sides by \\\$r\\\$ (valid for \\\$r \\neq 0\\\$)",
                    sol("Answer", polar_eq),
                ),
                time=150,
            )
        end
        
    elseif problem_type == :polar_curve_intersection
        # Find intersection points of two polar curves
        a = abs(nonzero(-8, 8))
        b = abs(nonzero(-8, 8))
        
        while a == b
            b = abs(nonzero(-8, 8))
        end
        
        # Two circles: r = a and r = b cos(θ)
        r1_eq = "r = $(a)"
        r2_eq = "r = $(2*b)\\cos(\\theta)"
        
        # Intersection: a = 2b cos(θ) → cos(θ) = a/(2b)
        if abs(a) <= abs(2*b)
            cos_val = a // (2*b)
            theta_val = "\\arccos($(cos_val))"
            num_points = 3  # (0,0) and two symmetric points
            
            problem(
                question="Find all intersection points of the polar curves \\\$$(r1_eq)\\\$ and \\\$$(r2_eq)\\\$. How many intersection points are there?",
                answer=num_points,
                difficulty=(2000, 2200),
                solution=steps(
                    "Given curves: \\\$$(r1_eq)\\\$ and \\\$$(r2_eq)\\\$",
                    "Set equal: \\\$$(a) = $(2*b)\\cos(\\theta)\\\$",
                    "Solve: \\\$\\cos(\\theta) = $(cos_val)\\\$",
                    "Also check pole: second curve passes through origin when \\\$\\theta = \\pi/2\\\$",
                    "Total intersection points: $(num_points) (including the pole)",
                ),
                time=240,
            )
        else
            problem(
                question="Find all intersection points of the polar curves \\\$$(r1_eq)\\\$ and \\\$$(r2_eq)\\\$. How many intersection points are there?",
                answer=1,
                difficulty=(2000, 2200),
                solution=steps(
                    "Given curves: \\\$$(r1_eq)\\\$ and \\\$$(r2_eq)\\\$",
                    "Set equal: \\\$$(a) = $(2*b)\\cos(\\theta)\\\$",
                    "This requires \\\$\\cos(\\theta) = $(a)/$(2*b) > 1\\\$, which is impossible",
                    "Check pole: second curve passes through origin at \\\$\\theta = \\pi/2\\\$",
                    "Only intersection point: the pole (origin)",
                ),
                time=240,
            )
        end
        
    elseif problem_type == :polar_area_setup
        # Set up integral for area enclosed by polar curve
        n = choice([3, 4, 5, 6, 8])
        a = abs(nonzero(-6, 6))
        
        curve_eq = "r = $(a)\\cos($(n)\\theta)"
        
        # Rose curve with n petals (if n odd) or 2n petals (if n even)
        num_petals = n % 2 == 0 ? 2*n : n
        area_one_petal = (a^2 * π) // (4 * n)
        total_area = (a^2 * π * num_petals) // (4 * n)
        
        problem(
            question="Find the total area enclosed by the rose curve \\\$$(curve_eq)\\\$.",
            answer=Float64(total_area),
            difficulty=(2100, 2300),
            solution=steps(
                "Given: \\\$$(curve_eq)\\\$",
                "This is a rose curve with $(num_petals) petals",
                "Area of one petal: \\\$A_{\\text{petal}} = \\frac{1}{2}\\int_{\\alpha}^{\\beta} r^2 \\, d\\theta\\\$",
                "For one petal: \\\$A_{\\text{petal}} = $(area_one_petal)\\\$",
                "Total area: \\\$A = $(num_petals) \\times $(area_one_petal) = $(total_area)\\\$",
            ),
            time=300,
        )
        
    elseif problem_type == :polar_distance
        # Distance between two polar points
        r1 = abs(randint(2, 12))
        r2 = abs(randint(2, 12))
        
        # Use nice angles
        angle_diff = choice([π//6, π//4, π//3, π//2, 2*π//3, 3*π//4, 5*π//6])
        
        # Distance formula: d² = r1² + r2² - 2r1·r2·cos(θ2 - θ1)
        dist_squared = r1^2 + r2^2 - 2*r1*r2*cos(Float64(angle_diff))
        dist = sqrt(dist_squared)
        
        angle_str = angle_diff == π//6 ? "\\frac{\\pi}{6}" :
                    angle_diff == π//4 ? "\\frac{\\pi}{4}" :
                    angle_diff == π//3 ? "\\frac{\\pi}{3}" :
                    angle_diff == π//2 ? "\\frac{\\pi}{2}" :
                    angle_diff == 2*π//3 ? "\\frac{2\\pi}{3}" :
                    angle_diff == 3*π//4 ? "\\frac{3\\pi}{4}" : "\\frac{5\\pi}{6}"
        
        problem(
            question="Find the distance between the points with polar coordinates \$($(r1), 0)\\\$ and \$($(r2), $(angle_str))\\\$.",
            answer=round(dist, digits=3),
            difficulty=(1900, 2100),
            solution=steps(
                "Given points: \$($(r1), 0)\\\$ and \$($(r2), $(angle_str))\\\$",
                "Use polar distance formula: \\\$d^2 = r_1^2 + r_2^2 - 2r_1 r_2 \\cos(\\theta_2 - \\theta_1)\\\$",
                "Substitute: \\\$d^2 = $(r1)^2 + $(r2)^2 - 2($(r1))($(r2))\\cos($(angle_str))\\\$",
                "Calculate: \\\$d^2 = $(round(dist_squared, digits=3))\\\$",
                sol("Answer", "d = $(round(dist, digits=3))"),
            ),
            time=180,
        )
        
    end
end
